### PR TITLE
Admin Panel

### PR DESCRIPTION
**Pull request by ritik-bluebash**
Issue:- #205 
Added admin panel.

https://user-images.githubusercontent.com/80153678/142415681-c6b0a12a-8580-4e87-8a69-ddc270d14beb.mp4

### CODE REVIEW

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+                                  params['order']['0']['dir'])
+    page_number = (params['start'].to_i / params['length'].to_i) + 1
+    users = users.page(page_number).per(params['length'])
+    render json: { data: users,
+                   draw: params['draw'].to_i,
+                   recordsTotal: User.count,
+                   recordsFiltered: filtered_users }
+  end
+
+  # def destroy
```
Comment (vishal-bluebash) [inline app/controllers/admin/users_controller.rb:19]: Let's remove these

File: app/datatables/admin/flags_datatable.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    def fetch_flags
+      search_string = []
+      columns.each do |term|
+        search_string << "#{term} like :search"
+      end
+
+      flags = Flag.order("#{sort_column} #{sort_direction}")
+      flags = flags.page(page).per(per_page)
+      flags = flags.where(search_string.join(' or '), search: "%#{params[:search][:value]}%")
```
Comment (vishal-bluebash) [inline app/datatables/admin/flags_datatable.rb:38]: @ritik-bluebash Let's use to wrap user input in `sanitize_sql`, check elsewhere as well we will not use user input directly in SQL queries

https://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html#method-i-sanitize_sql

File: app/javascript/packs/admin.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// import jQuery from 'jquery';
+// import Rails from 'rails-ujs';
+// import Turbolinks from 'turbolinks';
```
Comment (vishal-bluebash) [inline app/javascript/packs/admin.js:3]: Let's remove these commented code

File: app/javascript/packs/application.js [hunk 4]
```diff
@@ -46,6 +46,9 @@ Turbolinks.start();
 import 'jquery-ui-dist/jquery-ui';
 import 'bootstrap/dist/js/bootstrap';
 
+import DataTable from 'datatables/media/js/jquery.dataTables';
+window.DataTable = DataTable;
```
Comment (vishal-bluebash) [inline app/javascript/packs/application.js:50]: @ritik-bluebash Why do we need this here? as we are using datatables in admin layout only?

File: app/javascript/packs/application.js [hunk 5]
```diff
@@ -46,6 +46,9 @@ Turbolinks.start();
...
+
 $(function () {
   console.log('Hello World from Webpacker');
```
Comment (vishal-bluebash) [inline app/javascript/packs/application.js:53]: Let's remove this

File: app/javascript/src/admin/scripts/search/index.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import * as $ from 'jquery';
+
+export default (function () {
+  $('.search-toggle').on('click', e => {
```
Comment (vishal-bluebash) [inline app/javascript/src/admin/scripts/search/index.js:4]: use of jquery

File: app/views/admin/homes/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<div class='col-md-3'>
+  <div class="layers bd bgc-white p-20">
+    <div class="layer w-100 mB-10">
+      <h6 class="lh-1">Welcome to PubliCV Admin Panel</h6>
```
Comment (vishal-bluebash) [inline app/views/admin/homes/show.html.erb:4]: We need ti setup translations for admin panel as well (Separate yml file)

File: app/views/admin/users/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    </div>
+
+  </div>
+</div>
+<script type="text/javascript">
+  document.addEventListener("turbolinks:load", function(event) {
+    $('#dataTable').dataTable({
+      "scrollX": true,
+      "retrieve": true,
+      ajax: {
```
Comment (vishal-bluebash) [inline app/views/admin/users/index.html.erb:18]: use of ajax, we need to use javascript

File: config/routes.rb [hunk 9]
```diff
@@ -25,6 +25,12 @@ def self.matches?(request)
     confirmations: 'users/confirmations'
   }
 
+  namespace :admin do
+    resource :home, only: [:show]
+    resources :users, only: %i[index destroy]
```
Comment (vishal-bluebash) [inline config/routes.rb:30]: get 'users/records', to: 'users#records'

This can be added in nested  of users

**Review by vishal-bluebash — CHANGES_REQUESTED** [review]
Reviewed and added comments

**Comment by vishal-bluebash** [comment]
@ritik-bluebash Let's finish this PR and make it ready to merge

File: app/views/admin/users/index.html.erb (hunk 8 above)
Comment (ritik-bluebash) [reply app/views/admin/users/index.html.erb:18]: Datatable initialization is based on jQuery

File: app/javascript/packs/application.js (hunk 4 above)
Comment (ritik-bluebash) [reply app/javascript/packs/application.js:51]: Updated to admin layout.

**Review by vishal-bluebash — APPROVED** [review]
Beta version is ready

**Pull request merged**